[JSC] Reduce the number of @generatorResume() parameters by 2

shvaikalesh opened the pull request:
#### 9549797c340c80f64fe468853e8df3179b8c4f43
<pre>
[JSC] Reduce the number of @generatorResume() parameters by 2
<a href="https://bugs.webkit.org/show_bug.cgi?id=264128">https://bugs.webkit.org/show_bug.cgi?id=264128</a>
&lt;<a href="https://rdar.apple.com/problem/117885568">rdar://problem/117885568</a>&gt;

Reviewed by Yusuke Suzuki.

This change hoists @GeneratorStateCompleted check from @generatorResume() to its callers because
all three of them behave differently in that case, and defers retrieving @generatorFieldThis
up until calling into generator function&apos;s body.

This is unobservable refactor that paves the way for future yield* spec conformance changes like
<a href="https://bugs.webkit.org/show_bug.cgi?id=174756">https://bugs.webkit.org/show_bug.cgi?id=174756</a>, makes implementation of iterator helpers&apos; methods
a bit nicer, and was proven to be 0.5-1% progression on generator-heavy JetStream3 subtests.

Initially, @generatorResume() was designed this way to keep its size under 100, but since that time
maximumFunctionForCallInlineCandidateBytecodeCostForFTL was increased to 125, making this change
feasible (new size: 103).

* Source/JavaScriptCore/builtins/GeneratorPrototype.js:
(linkTimeConstant.generatorResume):
(next):
(return):
(throw):

Canonical link: <a href="https://commits.webkit.org/284511@main">https://commits.webkit.org/284511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97c66a0ede475751a0ecdd6574552e7ed405ed7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69670 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22421 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73755 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/20828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56871 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20679 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55370 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/20828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72736 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44736 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60092 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35849 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41402 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19205 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62786 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/63338 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17875 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75468 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68916 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13652 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17130 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63041 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13692 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60175 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62957 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10982 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90698 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10626 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44874 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16100 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45948 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
I think this is done to make `generatorResume` inlineable small size.
Can you ensure that `generatorResume` is still smaller than 120?

Constellation reviewed (approved):
r=me if it is neutral on A/B test (in particular JS2/Basic).

Commit:
[JSC] Reduce the number of @generatorResume() parameters by 2
https://bugs.webkit.org/show_bug.cgi?id=264128
<rdar://problem/117885568>

Reviewed by Yusuke Suzuki.

This change hoists @GeneratorStateCompleted check from @generatorResume() to its callers because
all three of them behave differently in that case, and defers retrieving @generatorFieldThis
up until calling into generator function's body.

This is unobservable refactor that paves the way for future yield* spec conformance changes like
https://bugs.webkit.org/show_bug.cgi?id=174756, makes implementation of iterator helpers' methods
a bit nicer, and was proven to be 0.5-1% progression on generator-heavy JetStream3 subtests.

Initially, @generatorResume() was designed this way to keep its size under 100, but since that time
maximumFunctionForCallInlineCandidateBytecodeCostForFTL was increased to 125, making this change
feasible (new size: 103).

* Source/JavaScriptCore/builtins/GeneratorPrototype.js:
(linkTimeConstant.generatorResume):
(next):
(return):
(throw):

Canonical link: https://commits.webkit.org/284511@main
